homeassistant/components/tplink_omada/update.py

```diff
@@ -0,0 +1,149 @@
+
+from homeassistant.components.update import UpdateEntity, UpdateEntityFeature
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.event import async_call_later
+
+from .const import DOMAIN
+from .controller import OmadaSiteController
+from .coordinator import OmadaCoordinator
+from .entity import OmadaDeviceEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class FirmwareUpdateStatus(NamedTuple):
+    """Firmware update information for Omada SDN devices."""
+
+    device: OmadaListDevice
+    firmware: OmadaFirmwareUpdate | None
+
+
+async def _get_firmware_updates(client: OmadaSiteClient) -> list[FirmwareUpdateStatus]:
+    devices = await client.get_devices()
+    return [
+        FirmwareUpdateStatus(
+            device=d,
+            firmware=None
+            if not d.need_upgrade
+            else await client.get_firmware_details(d),
+        )
+        for d in devices
+    ]
+
+
+async def _poll_firmware_updates(
+    client: OmadaSiteClient,
+) -> dict[str, FirmwareUpdateStatus]:
+    """Poll the state of Omada Devices firmware update availability."""
+    return {d.device.mac: d for d in await _get_firmware_updates(client)}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up switches."""
+    controller: OmadaSiteController = hass.data[DOMAIN][config_entry.entry_id]
+    omada_client = controller.omada_client
+
+    devices = await omada_client.get_devices()
+
+    coordinator = OmadaCoordinator[FirmwareUpdateStatus](
+        hass,
+        omada_client,
+        "Firmware Updates",
+        _poll_firmware_updates,
+        poll_delay=6 * 60 * 60,
+    )
+
+    entities: list = []
+    for device in devices:
+        entities.append(OmadaDeviceUpdate(coordinator, device))
+
+    async_add_entities(entities)
+    await coordinator.async_request_refresh()
+
+
+class OmadaDeviceUpdate(
+    OmadaDeviceEntity[FirmwareUpdateStatus],
+    UpdateEntity,
+):
+    """Firmware update status for Omada SDN devices."""
+
+    _attr_supported_features = (
+        UpdateEntityFeature.INSTALL
+        | UpdateEntityFeature.PROGRESS
+        | UpdateEntityFeature.RELEASE_NOTES
+    )
+    _firmware_update: OmadaFirmwareUpdate = None
+
+    def __init__(
+        self,
+        coordinator: OmadaCoordinator[FirmwareUpdateStatus],
+        device: OmadaListDevice,
+    ) -> None:
+        """Initialize the update entity."""
+        super().__init__(coordinator, device)
+
+        self._mac = device.mac
+        self._device = device
+        self._omada_client = coordinator.omada_client
+
+        self._attr_unique_id = f"{device.mac}_firmware"
+        self._attr_has_entity_name = True
+        self._attr_name = "Firmware Update"
```

Comment:
Our new entity naming policies require it to be sentence casing.
Please use `Firmware update` instead.